README.md
![nano API](nano.png)
# Nano

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]


Nano is a very very tiny php library that allows you to create very fast rest APIs. 

Think it's like Slim but if Slim is slim, Nano is anorexic.

## Requirements

Strictly requires PHP 7.4.

## Install

Via Composer

``` bash
$ composer require midorikocak/nano
```

## Usage

Simply instantiate and include in your app.

``` php
use midorikocak\nano\Api;

require __DIR__ . '/vendor/autoload.php';

$api = new Api();

```

I know. It's not static.

### Defining REST resources 

Defining rest routes and using wildcards are easy.

``` php
$message = 'Welcome to Nano';

$api->get('/', function () use ($message) {
    echo json_encode(['message' => $message]);
    http_response_code(200);
});

$api->post('/', function () use ($message) {
    $input = (array)json_decode(file_get_contents('php://input'), true, 512, JSON_THROW_ON_ERROR);
    echo json_encode($input);
    http_response_code(201);
});

$api->get('/echo/{$message}', function ($message) {
    echo json_encode(['message' => $message]);
    http_response_code(200);
});

```

### Basic Auth

It's possible hide your routed behind an authentication layer. Currently it expects basic auth, more methods to come soon.

``` php

$authFunction = function ($username, $password) {
    return ($username == 'username' && $password == 'password');
};

$api->auth(function () use (&$api) {

    $api->get('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id]);
        http_response_code(201);
    });

    $api->post('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id]);
        http_response_code(201);
    });

    $api->put('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id]);
        http_response_code(204);
    });

    $api->delete('/entries/{id}', function ($id) {
        http_response_code(204);
    });

}, $authFunction);
```

Hence the basic auth is not encrypted, using https is strictly advised.

## Testing

You can test your live API using `Guzzle/Client`

``` php
<?php

declare(strict_types=1);

namespace midorikocak\nano;

use GuzzleHttp\Client;
use PHPUnit\Framework\TestCase;

class IntegrationTest extends TestCase
{
    public function testGet(): void
    {
        $client = new Client(
            [
                'base_uri' => $this->baseUri,
                'http_errors' => false,
            ],
        );

        $response = $client->request('GET', '/echo/hello');
        $this->assertEquals(200, $response->getStatusCode());
        $this->assertStringContainsString('hello', (string)$response->getBody());
    }
```

## Motivation

Mostly educational purposes.

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email mtkocak@gmail.com instead of using the issue tracker.

## Credits

- [Midori Kocak][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/midorikocak/nano.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/midorikocak/nano/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/midorikocak/nano.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/midorikocak/nano.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/midorikocak/nano.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/midorikocak/nano
[link-travis]: https://travis-ci.org/midorikocak/nano
[link-scrutinizer]: https://scrutinizer-ci.com/g/midorikocak/nano/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/midorikocak/nano
[link-downloads]: https://packagist.org/packages/midorikocak/nano
[link-author]: https://github.com/midorikocak
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.1] - 2020-01-23

### Added
- Rest API
- Authenticator callback

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing

## [1.2.3] - 2020-02-29

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Conflicting endpoints with wildcards

### Removed
- Nothing

### Security
- Nothing

## [1.3.1] - 2020-03-04

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Fix optional wildcards

### Removed
- Nothing

### Security
- Nothing


tests/ApiUnitTest.php
<?php

declare(strict_types=1);

namespace midorikocak\nano;

use PHPUnit\Framework\TestCase;

use function http_response_code;
use function json_encode;

use const JSON_THROW_ON_ERROR;

final class ApiUnitTest extends TestCase
{
    private Api $api;

    public function setUp(): void
    {
        parent::setUp();
        $this->api = new Api();
    }

    public function tearDown(): void
    {
        parent::tearDown();
        unset($this->api);
    }

    /**
     * @runInSeparateProcess
     */
    public function testApi(): void
    {
        $api = $this->api;
        $message = 'Hello REST';
        $this->assertNull(
            $api->get('/', function () {
                http_response_code(200);
            })
        );
    }

    /**
     * @runInSeparateProcess
     */
    public function testWildcards(): void
    {
        $api = $this->api;
        $this->assertNull(
            $api->get('/echo/{$message}', function ($message) {
                echo json_encode(['message' => $message], JSON_THROW_ON_ERROR, 512);
                http_response_code(200);
            })
        );
    }

    /**
     * @runInSeparateProcess
     */
    public function testAuth(): void
    {
        $api = $this->api;
        $auth = function ($username, $password) {
            return true;
        };
        $this->assertNull(
            $api->auth(
                function () use ($api) {
                    $api->get('/echo/{$message}', function ($message) {
                        echo json_encode(['message' => $message], JSON_THROW_ON_ERROR, 512);
                        http_response_code(200);
                    });
                },
                $auth
            )
        );
    }

    /**
     * @runInSeparateProcess
     */
    public function testUnauthorized(): void
    {
        $api = $this->api;
        $auth = function ($username, $password) {
            return false;
        };
        $this->assertNull(
            $api->auth(
                function () use ($api) {
                    $api->get('/echo/{$message}', function ($message) {
                        echo json_encode(['message' => $message], JSON_THROW_ON_ERROR, 512);
                        http_response_code(200);
                    });
                },
                $auth
            )
        );
    }
}


tests/IntegrationTest.php
<?php

declare(strict_types=1);

namespace midorikocak\nano;

use GuzzleHttp\Client;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Process\Process;

use function explode;
use function usleep;

class IntegrationTest extends TestCase
{
    private Client $http;

    private $baseUri;

    private static Process $process;

    public function setUp(): void
    {
        $this->baseUri = 'localhost:8080';
        $this->http = new Client(
            [
                'base_uri' => $this->baseUri,
                'http_errors' => false,
            ],
        );
    }

    public function tearDown(): void
    {
        unset($this->http);
    }

    public static function setUpBeforeClass(): void
    {
        self::$process = new Process(explode(' ', 'php -S localhost:8080 -t .'));
        self::$process->start();

        usleep(100000); //wait for server to get going
    }

    public static function tearDownAfterClass(): void
    {
        self::$process->stop();
    }

    public function test404(): void
    {
        $response = $this->http->request('GET', '/not-found');
        $this->assertEquals(404, $response->getStatusCode());
    }

    public function test200(): void
    {
        $response = $this->http->request('GET', '/');
        $this->assertEquals(200, $response->getStatusCode());
    }

    public function testWildcard(): void
    {
        $response = $this->http->request('GET', '/echo/hello');
        $this->assertEquals(200, $response->getStatusCode());
        $this->assertStringContainsString('hello', (string) $response->getBody());
    }

    public function testGet(): void
    {
        $response = $this->http->request('GET', '/');

        $this->assertEquals(200, $response->getStatusCode());

        $contentType = $response->getHeaders()['Content-Type'][0];
        $this->assertStringContainsString('application/json', $contentType);
    }

    public function testPost(): void
    {
        $data = ['message' => 'Hello Server'];
        $response = $this->http->request('POST', '/entries/1', [
            'json' => $data,
            'http_errors' => false,
            'headers' => [
                'Authorization' => 'Basic dXNlcm5hbWU6cGFzc3dvcmQ=',
            ],
        ]);

        $this->assertEquals(201, $response->getStatusCode());
    }

    public function testUnauthorized(): void
    {
        $data = ['message' => 'Hello Server'];
        $response = $this->http->request('POST', '/entries/1', [
            'json' => $data,
            'http_errors' => false,
            'headers' => [
                'Authorization' => 'Basic ljknw=',
            ],
        ]);

        $this->assertEquals(401, $response->getStatusCode());
    }

    public function testDelete(): void
    {
        $response = $this->http->request('DELETE', '/entries/1', [
            'http_errors' => false,
            'headers' => [
                'Authorization' => 'Basic dXNlcm5hbWU6cGFzc3dvcmQ=',
            ],
        ]);

        $this->assertEquals(204, $response->getStatusCode());
    }

    public function testPut(): void
    {
        $response = $this->http->request('PUT', '/entries/1', [
            'auth' => [
                'username',
                'password',
            ],
        ]);

        $this->assertEquals(204, $response->getStatusCode());
    }
}


src/Api.php
<?php

declare(strict_types=1);

namespace midorikocak\nano;

use midorikocak\nano\Exceptions\NotFoundException;

use function array_key_exists;
use function array_map;
use function array_shift;
use function array_values;
use function base64_decode;
use function count;
use function explode;
use function header;
use function http_response_code;
use function is_callable;
use function is_string;
use function parse_url;
use function preg_match;
use function preg_match_all;
use function preg_replace;
use function strncasecmp;
use function strtolower;
use function substr;
use function trim;

use const PHP_URL_PATH;

class Api
{
    private array $endpoints = [];
    private array $wildcards = [];

    private string $origin = '*';
    private int $responseCode = 404;

    public function __construct($origin = '*')
    {
        $this->origin = $origin;
    }

    public function setPrefix(string $prefix): void
    {
        $prefixEndpoints = [];
        $prefix = trim($prefix, '/');

        foreach ($this->endpoints as $methodName => $item) {
            if (!isset($prefixEndpoints[$methodName])) {
                $prefixEndpoints[$methodName] = [];
            }

            foreach ($item as $key => $value) {
                $key = $key === '' ? $key : '/' . $key;
                $prefixEndpoints[$methodName][$prefix . $key] = $value;
            }
        }

        $this->endpoints = $prefixEndpoints;

        $prefixWildcards = array_map(
            static function ($item) use ($prefix) {
                return $prefix . '/' . $item;
            },
            $this->wildcards
        );

        $this->wildcards = $prefixWildcards;
    }

    private function isOptions(): bool
    {
        if ((($_SERVER['REQUEST_METHOD'] ?? 'GET') === 'OPTIONS') && $this->checkOrigin()) {
            header('Access-Control-Max-Age: 1728000');
            header('Content-Length: 0');
            header('Content-Type: text/plain');
            http_response_code(200);
            return true;
        }

        header('Access-Control-Max-Age: 3600');
        header(
            'Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With'
        );
        return false;
    }

    private function checkOrigin(): bool
    {
        if ($this->origin !== '*' && !$_SERVER['HTTP_ORIGIN'] === $this->origin) {
            header('HTTP/1.1 403 Access Forbidden');
            header('Content-Type: text/plain');
            return false;
        }
        return true;
    }

    public function __destruct()
    {
        header("Access-Control-Allow-Origin: $this->origin");
        header('Content-Type: application/json; charset=UTF-8');
        header('Access-Control-Allow-Methods: OPTIONS, POST, GET, PUT, DELETE');
        header('Access-Control-Allow-Headers: Authorization, Origin, X-Requested-With, Content-Type, Accept');

        $method = strtolower($_SERVER['REQUEST_METHOD'] ?? 'GET');
        if ($method === 'options') {
            $this->isOptions();
        } else {
            $uri = trim(parse_url($_SERVER['REQUEST_URI'] ?? '/', PHP_URL_PATH), '/');

            // Ignore uri that starts .php file extension
            //$uri = preg_replace('/^(.*?\.php\/{0,1})/', '', $uri);
            if (!isset($this->endpoints[$method])) {
                $this->endpoints[$method] = [];
            }
            $compared = $this->compareAgainstWildcards($uri);
            if (isset($this->endpoints[$method][$uri])) {
                $fn = $this->endpoints[$method][$uri];
                $this->responseCode = 200;
                $fn();
            } elseif (!empty($compared)) {
                if (!array_key_exists($compared['pattern'], $this->endpoints[$method])) {
                    throw new NotFoundException();
                }
                $fn = $this->endpoints[$method][$compared['pattern']];
                $this->responseCode = 200;
                $fn(...$compared['values']);
            }
        }

        if ($this->responseCode && ((string) http_response_code() === '200')) {
            http_response_code($this->responseCode);
        }
    }

    private function compareAgainstWildcards($uri): array
    {
        foreach ($this->wildcards as $wildcard) {
            $compareUri = $this->compareUri($uri, $wildcard);
            if (!empty($compareUri)) {
                return $compareUri;
            }
        }
        return [];
    }

    private function compareUri($uri, $pattern): array
    {
        // does url have brackets?
        $hasBrackets = preg_match_all('~{(.+)}~', $pattern, $vars);
        if ($hasBrackets) {
            $newPattern = preg_replace('~/*{.+?}~m', '/*([^/{}]*)', $pattern);
            $passesNewPattern = preg_match('~^' . $newPattern . '$~', $uri, $values);
            array_shift($values);
            if ($passesNewPattern) {
                return [
                    'pattern' => $pattern,
                    'uri' => $uri,
                    'vars' => $vars,
                    'values' => array_values($values),
                ];
            }
        }
        return [];
    }

    private function hasBrackets($uri): bool
    {
        return preg_match('/{(.*?)}/', $uri) !== false;
    }

    public function getResponseCode(int $code): void
    {
        $this->responseCode = $code;
    }

    public function auth(callable $fn, callable $login)
    {
        if (isset($_SERVER['HTTP_AUTHORIZATION']) && strncasecmp($_SERVER['HTTP_AUTHORIZATION'], 'basic ', 6) === 0) {
            $exploded = explode(':', base64_decode(substr($_SERVER['HTTP_AUTHORIZATION'], 6)), 2);

            if (count($exploded) === 2) {
                [$un, $pw] = $exploded;
            }
            if ($login($un ?? '', $pw ?? '')) {
                $fn();
            } else {
                $this->responseCode = 401;
            }
        }
        if (isset($_SESSION['user'])) {
            $fn();
        }
    }

    public function __call($name, $arguments)
    {
        if (!isset($this->endpoints[$name])) {
            $this->endpoints[$name] = [];
        }

        if (count($arguments) !== 2) {
            return;
        }

        if (is_string($arguments[0]) && is_callable($arguments[1])) {
            $endpoint = parse_url(trim($arguments[0], '/'), PHP_URL_PATH);
            if ($this->hasBrackets($arguments[0])) {
                $this->wildcards[] = $endpoint;
            }
            $this->endpoints[$name][$endpoint] = $arguments[1];
        }
    }
}


index.php
<?php

declare(strict_types=1);

use midorikocak\nano\Api;

require __DIR__ . '/vendor/autoload.php';

$api = new Api();


$message = 'Welcome to Nano';

$api->get('/', function () use ($message) {
    echo json_encode(['message' => $message], JSON_THROW_ON_ERROR, 512);
    http_response_code(200);
});

$api->post('/', function () use ($message) {
    $input = (array)json_decode(file_get_contents('php://input'), true, 512, JSON_THROW_ON_ERROR);
    echo json_encode($input, JSON_THROW_ON_ERROR, 512);
    http_response_code(201);
});

$api->get('/echo/{$message}', function ($message) {
    echo json_encode(['message' => $message], JSON_THROW_ON_ERROR, 512);
    http_response_code(200);
});


$authFunction = function ($username, $password) {
    return ($username == 'username' && $password == 'password');
};

$api->auth(function () use (&$api) {

    $api->get('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id], JSON_THROW_ON_ERROR, 512);
        http_response_code(201);
    });


    $api->post('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id], JSON_THROW_ON_ERROR, 512);
        http_response_code(201);
    });

    $api->put('/entries/{id}', function ($id) {
        echo json_encode(['id' => $id], JSON_THROW_ON_ERROR, 512);
        http_response_code(204);
    });

    $api->delete('/entries/{id}', function ($id) {
        http_response_code(204);
    });

}, $authFunction);
